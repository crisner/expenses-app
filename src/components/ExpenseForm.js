import 'react-dates/initialize';
import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        this.setState(() => ({ createdAt }));
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused }));
    };



    render() {
        return (
            <form>
                <input type="text" placeholder="Description" value={this.state.description} onChange={this.onDescriptionChange} autoFocus />
                <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
                <SingleDatePicker
                 date={this.state.createdAt}
                 onDateChange={this.onDateChange}
                 focused={this.state.focused}
                 onFocusChange={this.onFocusChange}
                 id="datepicker" />
                <textarea placeholder="Enter note(optional)" value={this.state.note} onChange={this.onNoteChange}></textarea>
                <button>Submit</button>
            </form>
        );
    }
}