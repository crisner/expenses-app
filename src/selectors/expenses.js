import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter(expense => {
        const startDateMatch = !startDate || moment(expense.createdAt).isSameOrAfter(startDate);
        const endDateMatch = !endDate || moment(expense.createdAt).isSameOrBefore(endDate);
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }

        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    })
};

export default getVisibleExpenses;