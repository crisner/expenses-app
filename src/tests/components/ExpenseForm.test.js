import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';

test('should render the expense form', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render the expense form with expense data', () => {
    const wrapper = shallow(<ExpenseForm {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error when submitting invalid form', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
    const value = 'New description';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('description')).toBe(value);
});

test('should set note on textarea change', () => {
    const value = 'New note';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('textarea').simulate('change', {
        target: { value }
    });
    expect(wrapper.state('note')).toBe(value);
});

test('should set amount if valid', () => {
    const value = '23.33';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount if input is invalid', () => {
    const value = '255.269';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ExpenseForm {...expenses[0]} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[0].description,
        note: expenses[0].note,
        amount: expenses[0].amount,
        createdAt: expenses[0].createdAt.valueOf()
    });
});

test('should set new date on date change', () => {
    const newDate = moment().toDate();
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('DayPickerInput').prop('onDayChange')(newDate);
    expect(wrapper.state('createdAt')).toBe(newDate);
});