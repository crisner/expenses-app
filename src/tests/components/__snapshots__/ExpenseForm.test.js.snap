// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error when submitting invalid form 1`] = `
<div>
  <div />
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DayPickerInput
      classNames={
        Object {
          "container": "DayPickerInput",
          "overlay": "DayPickerInput-Overlay",
          "overlayWrapper": "DayPickerInput-OverlayWrapper",
        }
      }
      clickUnselectsDay={false}
      component="input"
      dayPickerProps={
        Object {
          "createdAts": 1970-01-01T00:00:00.000Z,
        }
      }
      format="ll"
      formatDate={[Function]}
      hideOnDayClick={true}
      inputProps={Object {}}
      keepFocus={true}
      onDayChange={[Function]}
      overlayComponent={[Function]}
      parseDate={[Function]}
      placeholder="Jan 1, 1970"
      showOverlay={false}
      value={1970-01-01T00:00:00.000Z}
    />
    <textarea
      onChange={[Function]}
      placeholder="Enter note(optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render error when submitting invalid form 2`] = `
<div>
  <div>
    <p>
      Please add a description and an amount
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DayPickerInput
      classNames={
        Object {
          "container": "DayPickerInput",
          "overlay": "DayPickerInput-Overlay",
          "overlayWrapper": "DayPickerInput-OverlayWrapper",
        }
      }
      clickUnselectsDay={false}
      component="input"
      dayPickerProps={
        Object {
          "createdAts": 1970-01-01T00:00:00.000Z,
        }
      }
      format="ll"
      formatDate={[Function]}
      hideOnDayClick={true}
      inputProps={Object {}}
      keepFocus={true}
      onDayChange={[Function]}
      overlayComponent={[Function]}
      parseDate={[Function]}
      placeholder="Jan 1, 1970"
      showOverlay={false}
      value={1970-01-01T00:00:00.000Z}
    />
    <textarea
      onChange={[Function]}
      placeholder="Enter note(optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render the expense form 1`] = `
<div>
  <div />
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DayPickerInput
      classNames={
        Object {
          "container": "DayPickerInput",
          "overlay": "DayPickerInput-Overlay",
          "overlayWrapper": "DayPickerInput-OverlayWrapper",
        }
      }
      clickUnselectsDay={false}
      component="input"
      dayPickerProps={
        Object {
          "createdAts": 1970-01-01T00:00:00.000Z,
        }
      }
      format="ll"
      formatDate={[Function]}
      hideOnDayClick={true}
      inputProps={Object {}}
      keepFocus={true}
      onDayChange={[Function]}
      overlayComponent={[Function]}
      parseDate={[Function]}
      placeholder="Jan 1, 1970"
      showOverlay={false}
      value={1970-01-01T00:00:00.000Z}
    />
    <textarea
      onChange={[Function]}
      placeholder="Enter note(optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render the expense form with expense data 1`] = `
<div>
  <div />
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Downpayment"
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value="250"
    />
    <DayPickerInput
      classNames={
        Object {
          "container": "DayPickerInput",
          "overlay": "DayPickerInput-Overlay",
          "overlayWrapper": "DayPickerInput-OverlayWrapper",
        }
      }
      clickUnselectsDay={false}
      component="input"
      dayPickerProps={
        Object {
          "createdAts": 1969-12-27T00:00:00.000Z,
        }
      }
      format="ll"
      formatDate={[Function]}
      hideOnDayClick={true}
      inputProps={Object {}}
      keepFocus={true}
      onDayChange={[Function]}
      overlayComponent={[Function]}
      parseDate={[Function]}
      placeholder="Dec 27, 1969"
      showOverlay={false}
      value={1969-12-27T00:00:00.000Z}
    />
    <textarea
      onChange={[Function]}
      placeholder="Enter note(optional)"
      value="For renting the new apartment"
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;
